src/clj/puppetlabs/services/jruby/jruby_metrics_core.clj

```diff
@@ -140,10 +147,13 @@
   :- TimestampedReasonWithRequestInfo
   [instance :- TimestampedReason]
   (if-let [request (get-in instance [:reason :request])]
-    (assoc-in instance [:reason :request]
-              {:uri (:uri request)
-               :request-method (:request-method request)
-               :route-id (get-in request [:route-info :route-id])})
+    (if (nil? (schema/check ComidiRequest request))
+      ;; Schema check passed. We can assume this is a full Comidi request.
+      (assoc-in instance [:reason :request]
+                {:uri (:uri request)
+                 :request-method (:request-method request)
+                 :route-id (get-in request [:route-info :route-id])})
+      (assoc instance :reason ::unknown-http-request))
```

Comment:
The other option would be to make `:route-id` an optional key in the schema.

Reply:
So we say the instance that we pass in should be a [`TimestampedReason`](https://github.com/puppetlabs/puppetserver/pull/1691/files#diff-06e8dcaea6b81022d68725a975548198R65) which is `{:time Long :reason schema/Any}`. Does it make sense to tighten up the input to this method instead?

Reply:
Ah, the problem isn't the input schema. It's the output schema of `TimestampedReasonWithRequestInfo` which requires that any `:reason` that is a map [that contains a `:request` key](https://github.com/puppetlabs/puppetserver/pull/1691/files#diff-06e8dcaea6b81022d68725a975548198R77) shall have a [non-null :route-id](https://github.com/puppetlabs/puppetserver/pull/1691/files#diff-06e8dcaea6b81022d68725a975548198R71)

A regular Ring mock causes this output assertion to fail as there is no Comidi `:route-info` for the `:route-id` to be extracted from.

Reply:
I'm sorry, I was having a weird day when I reviewed these two PRs the first time. I meant to ask that comment wrt the schema check on line 150:
```clj
    (if (nil? (schema/check ComidiRequest request))
```
Should we, instead of adding more checking inside the method, tighten up the validation that schema is doing on the initial input? ie, if `TimestampedReason` contains the `:request` key that key should be a `ComidiRequest` (I assume something with the `:request` key should have some kind of valid request even if its not comidi)?

If not, ie if we need to support non-comidi requests, then for this section can we just give the `:route-id` a default value? Like:
```clj
                 :route-id (get-in request [:route-info :route-id] ::unknown-http-request)})
```

Reply:
I guess my concern with this function's changes are: is this doing work in our function body that schema should be doing for us in the input validation? and is it actually valid behavior to support non-comidi request objects outside of tests (old compojure or liberator maybe??)?

Reply:
> Should we, instead of adding more checking inside the method, tighten up the validation that schema is doing on the initial input? ie, if TimestampedReason contains the :request key that key should be a ComidiRequest (I assume something with the :request key should have some kind of valid request even if its not comidi)?

The `:request` key should always hold a Ring request map which can optionally include some Comidi metadata added by the routes in the `MasterService`. The output schema of the `instance-request-info` function currently requires that Comidi metadata be present.

If we tighten up the input validation, then a bunch of unit tests for the `RequestHandlerService` that currently mock Ring request maps will have to start mocking Ring+Comidi request maps or pull in a dependency on the `MasterServer` --- both of those options seem like bad directions to go as they increase the coupling of the code.

I like the idea of just adding a default to the `get-in` function that returns `"unknown-http-request"` for `:route-id` if the Comidi metadata isn't present. Seems like the most elegant solution that will satisfy the existing schemas.